pkg/controller/clusterpool/collections.go

```diff
@@ -584,6 +617,475 @@ func (cds *cdCollection) Delete(c client.Client, cdName string) error {
+	for _, item := range pool.Spec.Inventory {
+		if cdc, ok := cdcCol.namespace[item.Name]; ok {
+			cdcCol.byCDCName[item.Name] = cdc
+			if cdRef := cdc.Status.ClusterDeploymentRef; cdRef == nil {
+				cdcCol.unassigned = append(cdcCol.unassigned, cdc)
+			} else {
+				cdcCol.reserved[item.Name] = cdc
+			}
+			applyStatus := conditionsv1.FindStatusCondition(cdc.Status.Conditions, hivev1.ApplySucceededCondition)
+			if applyStatus == nil {
+				continue
+			}
+			if applyStatus.Reason == hivev1.CustomizationApplyReasonBrokenCloud {
+				cdcCol.cloud[item.Name] = cdc
+			}
+			if applyStatus.Reason == hivev1.CustomizationApplyReasonBrokenSyntax {
+				cdcCol.syntax[item.Name] = cdc
+			}
+		} else {
+			cdcCol.missing = append(cdcCol.missing, item.Name)
+		}
+	}
+
+	cdcCol.Sort()
+
+	logger.WithFields(log.Fields{
+		"unassignedCount":     len(cdcCol.unassigned),
+		"missingCount":        len(cdcCol.missing),
+		"reservedCount":       len(cdcCol.reserved),
+		"brokenByCloudCount":  len(cdcCol.cloud),
+		"brokenBySyntaxCount": len(cdcCol.syntax),
+	}).Debug("found ClusterDeploymentCustomizations for ClusterPool")
+
+	return &cdcCol, nil
+}
+
+// Sort unassigned oldest successful customizations to avoid using the same broken
+// customization.  When customizations have the same last apply status, the
+// oldest used customization will be prioritized.
+func (cdcs *cdcCollection) Sort() {
+	sort.Slice(
+		cdcs.unassigned,
+		func(i, j int) bool {
+			now := metav1.NewTime(time.Now())
+			iStatus := conditionsv1.FindStatusCondition(cdcs.unassigned[i].Status.Conditions, hivev1.ApplySucceededCondition)
+			jStatus := conditionsv1.FindStatusCondition(cdcs.unassigned[j].Status.Conditions, hivev1.ApplySucceededCondition)
+			iName := cdcs.unassigned[i].Name
+			jName := cdcs.unassigned[j].Name
+			if iStatus == nil || iStatus.Status == corev1.ConditionUnknown {
+				iStatus = &conditionsv1.Condition{Reason: hivev1.CustomizationApplyReasonSucceeded}
+				iStatus.LastTransitionTime = now
+			}
+			if jStatus == nil || jStatus.Status == corev1.ConditionUnknown {
+				jStatus = &conditionsv1.Condition{Reason: hivev1.CustomizationApplyReasonSucceeded}
+				jStatus.LastTransitionTime = now
+			}
+			iTime := iStatus.LastTransitionTime
+			jTime := jStatus.LastTransitionTime
+			if iStatus.Reason == jStatus.Reason {
+				if iTime.Equal(&jTime) {
+					// Sort by name to make this deterministic
+					return iName < jName
+				}
+				return iTime.Before(&jTime)
+			}
+			if iStatus.Reason == hivev1.CustomizationApplyReasonSucceeded {
+				return true
+			}
+			if jStatus.Reason == hivev1.CustomizationApplyReasonSucceeded {
+				return false
+			}
+			return iName < jName
+		},
+	)
+}
+
+func (cdcs *cdcCollection) Reserve(c client.Client, cdc *hivev1.ClusterDeploymentCustomization, cdName, poolName string) error {
+	cdc.Status.ClusterDeploymentRef = &corev1.LocalObjectReference{Name: cdName}
+	cdc.Status.ClusterPoolRef = &corev1.LocalObjectReference{Name: poolName}
```

Comment:
We're assuming these are changing, but we're checking the status condition below. We should either be checking all, or providing a good justification (in a comment) for why we don't need to.